feat(gateway): visualize dag-cbor and dag-json

hacdias opened the pull request:
Closes #307. Builds on top of #298.

- Adds minimal DAG preview for DAG-CBOR and DAG-JSON
- You can preview how it looks by following `gateway/assets/README.md`
	- For live version, clone Kubo and this PR and then `go mod edit -replace=github.com/ipfs/boxo=path/to/your/local/boxo`
- Sharness fails here due to a bug fix: see https://github.com/ipfs/kubo/pull/9904.

<img width="1340" alt="image" src="https://github.com/ipfs/boxo/assets/5447088/227829ef-b89f-416d-a426-a231f071e130">

<img width="1340" alt="image" src="https://github.com/ipfs/boxo/assets/5447088/2cde8e3e-66cf-4031-8891-0785da257246">

hacdias commented:
I'm opening this for review to get feedback and for others to take it for a spin. I was quite conservative on how we handle the preview. If some error happens, no preview is shown (this usually means that the DAG isn't valid). If it all works, we show a preview.

lidel reviewed (changes requested):
Thank you @hacdias I think this is good enough to ship.

Remaining work:
-  [ ] We need Kubo PR testing this after https://github.com/ipfs/boxo/pull/298#pullrequestreview-1449906423 lands
- [ ] Add a regression test that confirms CBOR bytes that get turned into strings in HTML get sanitized (just to be sure)

hacdias commented on gateway/assets/node.go:
```diff
@@ -0,0 +1,110 @@
+				return nil, err
+			}
+
+			dag.Keys = append(dag.Keys, &ParsedNode{Value: fmt.Sprint(k)})
```
@lidel `k` is an `int64`. The only one I wrapped with `template.HTMLEscapeString` is a string. The remaining ones are safe: booleans, integers, floats, cids, etc. The bytes will be printed as `[ 000 000 000 000 ]`, or whatever is the hex. So I think sanitisation is correct. I will add a test to be sure.

hacdias replied:
We may also consider using [`hex.Dump`](https://pkg.go.dev/encoding/hex#Dump) instead.

lidel replied:
@hacdias ooh, `hex.Dump` would be significantly better: mirrors `hexdump -C` and removes the need for someone to download raw block -- can inspect both HEX and ASCII representation right in the browser.

https://play.golang.org/p/jj5itycHpKi

Could we swithc to it and put the produced string in `<pre>` tag?

hacdias replied:
Done.

hacdias commented on gateway/assets/node.go:
```diff
@@ -0,0 +1,110 @@
+		if err != nil {
+			return nil, err
+		}
+		dag.Value = fmt.Sprint(v)
```
See https://github.com/ipfs/boxo/pull/315#discussion_r1210051985.

hacdias commented:
Kubo is green: https://github.com/ipfs/kubo/pull/9904

lidel reviewed (changes requested):
@hacdias while testing in Kubo, I've found some things we should address before merging:
- [ ]  subdomain gateways should never return dag.html  from a path, only return redirect: http://localhost:8080/ipfs/bafyreihnpl7ami7esahkfdnemm6idx4r2n6u3apmtcrxlqwuapgjsciihy MUST  redirect to http://bafyreihnpl7ami7esahkfdnemm6idx4r2n6u3apmtcrxlqwuapgjsciihy.ipfs.localhost:8080
- [ ] while at this, for HTML responses, we should also ensure the generated  HTML document is returned with a scope limited to the CID, the same way dir listings are. On path gateway http://127.0.0.1:8080/ipfs/bafyreihnpl7ami7esahkfdnemm6idx4r2n6u3apmtcrxlqwuapgjsciihy MUST return HTTP 301 redirect to http://127.0.0.1:8080/ipfs/bafyreihnpl7ami7esahkfdnemm6idx4r2n6u3apmtcrxlqwuapgjsciihy/ (normalize with `/` at the end, just like unixfs dirs)
  - this is an important precaution in case there is ever a bug in sanitization and/or our [service worker protection is ever removed](https://github.com/ipfs/boxo/blob/c23df3837fbe224259a60e8eed9d77e2656f2dce/gateway/handler.go#L803-L810)

hacdias commented:
@lidel I just pushed a fix for both, including tests. In addition, I have a few notes:

- The codec handler was never redirecting properly. I changed the converted and the raw to wrap the `ResponseWriter` with `statusResponseWriter` so we preserve the redirection status. Similar to UnixFS files.
- HTML response now redirects to `/`. Therefore, some sharness tests in Kubo fail where we used to check if we got HTML back. I fixed them by simply requesting the path with trailing slash. The redirection is tested in the new tests in this PR. https://github.com/ipfs/kubo/pull/9904

hacdias commented:
@lidel regarding our redirects and body conversations, I did not include it here because we have some tests currently checking for that: https://github.com/ipfs/boxo/pull/326

lidel reviewed (approved):
Thank you, lgtm, merging. I will bubble up to https://github.com/ipfs/kubo/pull/9904 and review there after CI is green.
